docs/design.md

```diff
@@ -0,0 +1,279 @@
+[Filecoin's Block Sync protocol](https://spec.filecoin.io/#section-algorithms.block_sync). The underlaying idea is the same, to "sync a chain of data".
+
+Thus, when calling `Sync(ctx, p, latest)`, indexer node sends a first request for ingestion as:
+```go
+req = IngestionRequest{
+        ID: latest,
+        Start: 0, 
+}
+```
+
+Provider will answer with the list of entries for `latest` pointing to the previous `AdvertisementID` in the `Advertisement` field. `Size` specifies
+the total number of entries that the advertisement include to give the indexer knowledge of when all entries for the update have been received.
+
+```go
+resp = IngestionResponse {
+        Size: 2
+        Error: 0
+        Start: 0
+        Advertisement: Advertisement{
+                ID: latest,
+                Previous: "latest-1",
+                IndexerID: Link_Index,
+                Provider: p,
+                Signature: []byte{},  // The original signature of the advertisement.
+                GraphSupport: false, 
+        }
+        Entries: {
+                []Entry{
+                        Cids: {Cid1, Cid2, ...},
+                        Metadata: []byte{},
+                }
+}
+```
+- If `Size > len(Entries)`, the indexer-node sends a new request to keep fetching data for `latest` with `Start: 1`
+until all entries for the Advertisement have been received. It then inspects `Advertisement` to see if `Previous` is equal
+to the latest `AdvertisementID` seen for that provider, if this is not the case, it sends a new request with `latest-1`
+and repeats the process over and over again until the `AdvertisementID` seen in `Previous` for the latest response
+equals the latest one seen by indexer node for that data provider and the full sync for the missing range is complete.
+
+Note that pagination is performed at an `Entry` level, i.e. there is no intra-Entry pagination (at least not initially). The data provider should be responsible for dividing single Entries with a large number of CIDs into
+smaller entries when computing the `Size` of the data indexed in `AdvertisementID`. If we identify this as too strong
+of a requirement, we can also introduce intra-Entry pagination to support seamless exchange of entries with a large
+amount of CIDs.
+
+Once all the data for an indexer has been fetched, the indexer can use the `Advertisement` and the list of entries received to authenticate the
+ingestion by computing the Cid of the entries and verifying that it is equal to `IndexID`, and then verifying the signature
+of the `Advertisement`.
+
+If the full sync is performed successfully, the indexer node updates the latest AdvertisementID seen for `p` to latest. 
+
+Indexer nodes run ingestion sessions for different providers in parallel, so an indexer node can sync with several provideres in parallel.
+They can also set a `prov_parallel_ingestion` config to determine if it wants to support parallel ingestions for
```

Comment:
I do not think we need `prov_parallel_ingestion`.  Since the diffs of a single provider need to be applied in order, it would mean that the indexer will have to temporarily store all the data for all updates except the oldest one, and this is not something we should do.

Reply:
You are right, parallel ingestion probably works for subsequents adds, but the moment an update arrives for the same CID in the wrong order things break (this would make sense if we used CRDTs which is not the case). I'lll remove this part.